Add RectangleIterator

AlexReimann opened the pull request:
Adds `RectangleIterator` class which is higher performance than the default grid map `PolygonIterator`.

It actually currently would also work with Polygons (if the size check at construction is removed).

Also changed the compiler options to use and enforce C++14

Commit:
Switch to C++14

Commit:
Add RectangleIterator

StefanFabian commented:
Maybe this might be interesting to you as well:
https://github.com/tu-darmstadt-ros-pkg/hector_math/blob/master/hector_math/include/hector_math/iterators/polygon_iterator.h

There's also a rectangle iterator but I just noticed I forgot to finish the implementation.
It has a different structure than the eth iterators but is easy to use with lambda functions and an order of magnitude faster.

AlexReimann commented:
> Maybe this might be interesting to you as well: https://github.com/tu-darmstadt-ros-pkg/hector_math/blob/master/hector_math/include/hector_math/iterators/polygon_iterator.h
> 

That repository is not accessible to me (or does not exist)

StefanFabian commented:
Oh, you're right. I'm sorry, looks like it isn't ready to be made public yet.

StefanFabian commented on src/rectangle_iterator.cpp:
```diff
@@ -0,0 +1,130 @@
+  };
+
+  start_index_ = get_index(max_x, max_y);
+  end_index_ = get_index(min_x, min_y);
```
Why is end min_x/y if you increment in nextStep?

AlexReimann replied:
Position coordinates and index directions are flipped:
https://github.com/ANYbotics/grid_map/blob/master/grid_map_core/doc/grid_map_conventions.pdf

I added a comment

Commit:
Fix RectangleIterator initial index

StefanFabian commented:
Three things are still unclear to me.
1. Why is this faster than the PolygonIterator if it uses the inBounds check of PolygonIterator and does some additional stuff?
2. This does not enforce right angles and hence is actually a QuadrangleIterator not a RectangleIterator. 
3. Why do you make a PR with this here instead of the grid_map repo where the other iterators are also located?

AlexReimann commented:
> Three things are still unclear to me.
> 
>     1. Why is this faster than the PolygonIterator if it uses the inBounds check of PolygonIterator and does some additional stuff?

It doesn't use the inBounds check of the PolygonIterator, which makes it faster. It skips checks on if the cell indices are in bounds, as this is already limited by checking for the min / max indices, which are calculated during construction.

We got a big speed increase also during tests and I'm using this, else I'd just drop it.


>     2. This does not enforce right angles and hence is actually a QuadrangleIterator not a RectangleIterator.

True that, but the plan was to change that to improve even more on the speed, by switching out the `isInside()` call on the polygon. Actually working on this today.
I can rename it for now if it makes you happy though.

>     3. Why do you make a PR with this here instead of the grid_map repo where the other iterators are also located?

The last time I had a 1 line fix in that repo and it took 2+ years to get it merged there and the packages released.
Also I asked @skohlbr if it's fine in here.

Commit:
Fix typo in RectangleIterator point count fatal msg

Commit:
Add missing Eigen include and dependency

Commit:
Replace RectangleIterator polygon isInside call

AlexReimann commented:
Replaced the `polygon_.isInside()` call with transforming the point into the rectangle coordinate system and doing simple x  / y checks.

Also added missing `Eigen` dependencies (now using more `Eigen`, but already was missing for the usage of `EIGEN_MAKE_ALIGNED_OPERATOR_NEW`)

AlexReimann commented:
Bump
